src/parser/parsers.h

```diff
@@ -895,6 +902,139 @@ template<typename Ctx> MaybeResult<> loop(Ctx& ctx, bool folded) {
   return ctx.visitEnd();
 }
 
+// trycatch ::= 'try' label blocktype instr* ('catch' id? tagidx instr*)*
+//                  ('catch_all' id? instr*)? 'end' id?
+//            | '(' 'try' label blocktype '(' 'do' instr* ')'
+//                  ('(' 'catch' tagidx instr* ')')*
+//                  ('(' 'catch_all' instr* ')')? ')'
+//            | 'try' label blocktype instr* 'deledate' label
+//            | '(' 'try' label blocktype '(' 'do' instr* ')'
+//                '(' 'delegate' label ')' ')'
+template<typename Ctx> MaybeResult<> trycatch(Ctx& ctx, bool folded) {
+  auto pos = ctx.in.getPos();
+
+  if ((folded && !ctx.in.takeSExprStart("try"sv)) ||
+      (!folded && !ctx.in.takeKeyword("try"sv))) {
+    return {};
+  }
+
+  auto label = ctx.in.takeID();
+
+  auto type = blocktype(ctx);
+  CHECK_ERR(type);
+
+  CHECK_ERR(ctx.makeTry(pos, label, *type));
+
+  if (folded) {
+    if (!ctx.in.takeSExprStart("do"sv)) {
+      return ctx.in.err("expected 'do' in try");
+    }
+  }
+
+  CHECK_ERR(instrs(ctx));
+
+  if (folded) {
+    if (!ctx.in.takeRParen()) {
+      return ctx.in.err("expected ')' at end of do");
+    }
+  }
+
+  if ((folded && ctx.in.takeSExprStart("delegate")) ||
+      (!folded && ctx.in.takeKeyword("delegate"))) {
+    auto delegatePos = ctx.in.getPos();
+
+    auto label = labelidx(ctx, true);
+    CHECK_ERR(label);
+
+    CHECK_ERR(ctx.visitDelegate(delegatePos, *label));
+    return Ok{};
+  }
+
+  while (true) {
+    auto catchPos = ctx.in.getPos();
+
+    if ((folded && !ctx.in.takeSExprStart("catch"sv)) ||
+        (!folded && !ctx.in.takeKeyword("catch"sv))) {
+      break;
+    }
+
+    // It can be ambiguous whether the name after `catch` is intended to be the
+    // optional ID or the tag identifier. First try parsing with the optional
+    // ID, then retry without parsing the ID if we run into an error.
+    bool parseID = !folded;
+    auto afterCatchPos = ctx.in.getPos();
+    while (true) {
+      if (!folded && parseID) {
+        auto id = ctx.in.takeID();
+        if (id && id != label) {
+          // Instead of returning an error, retry without the ID.
+          parseID = false;
+          ctx.in.lexer.setIndex(afterCatchPos);
+          continue;
+        }
+      }
+
+      auto tag = tagidx(ctx);
+      if (parseID && tag.getErr()) {
+        // Instead of returning an error, retry without the ID.
+        parseID = false;
+        ctx.in.lexer.setIndex(afterCatchPos);
+        continue;
+      }
```

Comment:
You're right that lines 967-975 parse the ID. This logic here tries to parse a tag (on line 977) and if that fails, it resets the parser to try again without trying to parse an ID.

Reply:
If a tag parsing fails, why do we set `parseID` false instead?
And what does it retry to do? Given that we set `parseID` false here, it doesn't retry to parse an ID. Then does it retry to parse a tag? But tag parsing just failed now.

Reply:
The ambiguous case this is handling looks like this:

```
(tag $t)
(func $ambiguous
  try $t
  catch $t
  end
)
```

When parsing the `catch`, the parser first tries to parse an optional ID that must match the label of the `try`, and it succeeds because it sees `$t` after the catch. However, when it then tries to parse the mandatory tag index, it fails because the next token is `end`. The problem is that the `$t` after the `catch` was the tag name and there was no optional ID after all. The parser sets `parseID = false` and resets to just after the `catch`, and now it skips parsing the optional ID so it correctly parses the `$t` as a tag name.

Reply:
Ah I see. Can you add this explanation in the comments? I think it will help reading.

Reply:
Done 👍